numpy/lib/function_base.py

```diff
@@ -2267,18 +2269,44 @@ class vectorize:
            [0., 0., 1., 2., 1., 0.],
            [0., 0., 0., 1., 2., 1.]])
 
+    Decorator syntax is supported.  The decorator can be called as
+    a function to provide keyword arguments.
+    >>>@np.vectorize
+    ...def identity(x):
+    ...    return x
+    ...
+    >>>identity([0, 1, 2])
+    array([0, 1, 2])
+    >>>@np.vectorize(otypes=[float])
+    ...def as_float(x):
+    ...    return x
+    ...
+    >>>as_float([0, 1, 2])
+    array([0., 1., 2.])
     """
-    def __init__(self, pyfunc, otypes=None, doc=None, excluded=None,
-                 cache=False, signature=None):
+    def __init__(self, pyfunc=np._NoValue, otypes=None, doc=None,
+                 excluded=None, cache=False, signature=None):
+
+        if (pyfunc != np._NoValue) and (not callable(pyfunc)):
```

Comment:
@mattip - I think is actually fairly safe for the decorator - if one passes in positional arguments, it is not very likely that the one for `otypes` is going to be callable.

That said, one could make it keyword only and deprecate the positional version, something like
```
def __init__(pyfunc=None, *args, otypes=None, ...):
    if args:
        warn(deprecation-warning)
        kwargs = {name : value
                  for (name, value, default) in [('otypes', otypes, None),
                                                 ...]
                  if value is not default}
        self._init_with_posargs(pyfunc, *args, **kwargs)

def _init_with_posargs(pyfunc=None, otypes=None, doc=None,
                 excluded=None, cache=False, signature=None):
    self._init(pyfunc, otypes=otypes, doc=doc, ...)
```